spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindImproperSynchronization.java

```diff
@@ -0,0 +1,807 @@
+            }
+        }
+
+        for (XField lock : inheritedCollectionLockObjects.keySet()) {
+            for (XField backingCollection : backingCollections.get(lock)) {
+                if (declaredCollectionAccessors.containsKey(backingCollection)) {
+                    reportAccessibleBackingCollectionBug(lock, inheritedCollectionLockObjects, backingCollection,
+                            declaredCollectionAccessors);
+                } else if (collectionAccessorsInHierarchy.containsKey(backingCollection)) {
+                    reportAccessibleBackingCollectionBug(lock, inheritedCollectionLockObjects, backingCollection,
+                            collectionAccessorsInHierarchy);
+                } else if (badBackingCollections.contains(backingCollection)) {
+                    reportBadCollectionObjectBugs(lock, declaredCollectionLockObjects, backingCollection);
+                }
+            }
+        }
+
+        clearState();
+    }
+
+    private boolean isCollection(Type type) {
+        if (type instanceof ReferenceType) {
+            ReferenceType rType = (ReferenceType) type;
+            try {
+                return Subtypes2.isContainer(rType);
+            } catch (ClassNotFoundException e) {
+                AnalysisContext.reportMissingClass(e);
+            }
+        }
+        return false;
+    }
+
+    private boolean isCollection(XField field) {
+        return isCollection(TypeFrameModelingVisitor.getType(field));
+    }
+
+    private boolean isMethodCall(Instruction instruction) {
+        return (instruction instanceof INVOKESTATIC)
+                || (instruction instanceof INVOKESPECIAL)
+                || (instruction instanceof INVOKEVIRTUAL)
+                || (instruction instanceof INVOKEINTERFACE);
+    }
+
+    private boolean isPackagePrivate(XField field) {
+        return !(field.isPublic() || field.isProtected() || field.isPrivate());
+    }
+
+    private boolean isInitializerMethod(String methodName) {
+        return Const.CONSTRUCTOR_NAME.equals(methodName) || Const.STATIC_INITIALIZER_NAME.equals(methodName);
+    }
+
+    /**
+     * Check if the lock object is inherited from the current class.
+     * To ensure no false positives, we compare only the outer class name if the lock object is contained in an inner class.
+     * Only checking equality is enough, since the lock object is a field declared in a class. That means if
+     * the declaring class of the lock object is the same as the current class, the lock object is not inherited.
+     * @param lockObject the lock object to check for inheritance
+     * @return true if the lock object is inherited from the current class, false otherwise
+     * @throws ClassNotFoundException if the class of the lock object cannot be found
+     */
+    private boolean isInherited(XField lockObject) throws ClassNotFoundException {
+        return !getDottedClassName().equals(lockObject.getClassName().split("\\$")[0]);
+    }
```

Comment:
`ClassNotFoundException` cannot be thrown from this method implementation. Can you please remove it from the signature and the javadoc? Also, if I'm not mistaken, this may have an impact on the callsites as well (the try blocks become unnecessary or at least the body of them can be minimized).

Reply:
You are right, this exception cannot be thrown from the function.
Generally, the callsites would also be affected, however, in these cases, other functions could throw the same exception.

Reply:
Yes, they can be simplified.